datacube/api/grid_workflow.py

```diff
@@ -0,0 +1,459 @@
+# This file is part of the Open Data Cube, see https://opendatacube.org for more information
+#
+# Copyright (c) 2015-2025 ODC Contributors
+# SPDX-License-Identifier: Apache-2.0
+from __future__ import annotations
+from collections.abc import Hashable
+import logging
+import typing
+import xarray as xr
+from collections import OrderedDict
+import pandas as pd
+
+from odc.geo.geom import intersects
+from .query import Query, query_group_by
+from .core import Datacube
+
+_LOG = logging.getLogger(__name__)
+
+if typing.TYPE_CHECKING:
+    from odc.geo.geobox import GeoBox
+    from datacube.model import Product
+    from datacube.index import Index
+
+
+def _fast_slice(array: xr.DataArray, indexers):
+    data = array.values[indexers]
+    dims = [
+        dim for dim, indexer in zip(array.dims, indexers) if isinstance(indexer, slice)
+    ]
+    coords = OrderedDict(
+        (
+            dim,
+            xr.Variable(
+                (dim,),
+                array.coords[dim].values[indexer],
+                attrs=array.coords[dim].attrs,
+                fastpath=True,
+            ),
+        )
+        for dim, indexer in zip(array.dims, indexers)
+        if isinstance(indexer, slice)
+    )
+    return xr.DataArray(data, dims=dims, coords=coords, attrs=array.attrs)
+
+
+class Tile:
+    """
+    The Tile object holds a lightweight representation of a datacube result.
+
+    It is produced by :meth:`.GridWorkflow.list_cells` or :meth:`.GridWorkflow.list_tiles`.
+
+    The Tile object can be passed to :meth:`GridWorkflow.load` to be loaded into memory as
+    an :class:`xr.Dataset`.
+
+    A portion of a tile can be created by using index notation. eg:
+
+        tile[0:1, 0:1000, 0:1000]
+
+    This can be used to load small portions of data into memory, instead of having to access
+    the entire `Tile` at once.
+    """
+
+    def __init__(self, sources: xr.DataArray, geobox):
+        """Create a Tile representing a dataset that can be loaded.
+
+        :param xr.DataArray sources: An array of non-spatial dimensions of the request, holding lists of
+            datacube.storage.DatasetSource objects.
+        :param model.GeoBox geobox: The spatial footprint of the Tile
+        """
+        self.sources: xr.DataArray = sources
+        self.geobox: GeoBox = geobox
+
+    @property
+    def dims(self) -> tuple[Hashable, ...]:
+        """Names of the dimensions, eg ``('time', 'y', 'x')``
+        :return: tuple(str)
+        """
+        return self.sources.dims + self.geobox.dimensions
+
+    @property
+    def shape(self) -> tuple[int, ...]:
+        """Lengths of each dimension, eg ``(285, 4000, 4000)``
+        :return: tuple(int)
+        """
+        return self.sources.shape + self.geobox.shape
+
+    @property
+    def product(self) -> Product:
+        """
+        :rtype: datacube.model.DatasetType
+        """
+        return self.sources.values[0][0].product
```

Comment:
Continuing typing note from above.  E.g. a DatasetSource would not have a product attribute, but a Dataset would.